ompi/mca/coll/libnbc/nbc_ireduce.c

```diff
@@ -285,65 +291,73 @@ int ompi_coll_libnbc_ireduce_inter(const void* sendbuf, void* recvbuf, int count
   if (vrank == 0) rank = root; \
   if (vrank == root) rank = 0; \
 }
-static inline int red_sched_binomial (int rank, int p, int root, const void *sendbuf, void *recvbuf, int count, MPI_Datatype datatype,
-                                      MPI_Op op, void *redbuf, NBC_Schedule *schedule, NBC_Handle *handle) {
-  int vrank, vpeer, peer, res, maxr;
-
-  RANK2VRANK(rank, vrank, root);
+static inline int red_sched_binomial (int rank, int p, int root, const void *sendbuf, void *redbuf, int count, MPI_Datatype datatype,
+                                      MPI_Op op, NBC_Schedule *schedule, NBC_Handle *handle) {
+  int vroot, vrank, vpeer, peer, res, maxr;
+  char *rbuf, *lbuf, *buf;
+  int tmprbuf, tmplbuf;
+
+  if (ompi_op_is_commute(op)) {
+    vroot = root;
+  } else {
+    vroot = 0;
+  }
+  RANK2VRANK(rank, vrank, vroot);
   maxr = (int)ceil((log((double)p)/LOG2));
 
+  /* ensure the result ends up in redbuf on vrank 0 */
+  if (0 == (maxr%2)) {
+    rbuf = 0;
+    tmprbuf = true;
+    lbuf = redbuf;
+    tmplbuf = false;
+  } else {
+    lbuf = 0;
+    tmplbuf = true;
+    rbuf = redbuf;
+    tmprbuf = false;
+  }
+
   for (int r = 1, firstred = 1 ; r <= maxr ; ++r) {
     if ((vrank % (1 << r)) == 0) {
       /* we have to receive this round */
       vpeer = vrank + (1 << (r - 1));
-      VRANK2RANK(peer, vpeer, root)
+      VRANK2RANK(peer, vpeer, vroot)
       if (peer < p) {
         /* we have to wait until we have the data */
-        res = NBC_Sched_recv (0, true, count, datatype, peer, schedule, true);
+        res = NBC_Sched_recv (rbuf, tmprbuf, count, datatype, peer, schedule, true);
         if (OPAL_UNLIKELY(OMPI_SUCCESS != res)) {
           return res;
         }
 
         /* perform the reduce in my local buffer */
         /* this cannot be done until handle->tmpbuf is unused :-( so barrier after the op */
-        if (firstred) {
-          if (rank == root) {
-            /* root is the only one who reduces in the receivebuffer
-             * take data from sendbuf in first round - save copy */
-            res = NBC_Sched_op (recvbuf, false, sendbuf, false, 0, true, count, datatype, op, schedule, true);
-          } else {
-            /* all others may not have a receive buffer
-             * take data from sendbuf in first round - save copy */
-            res = NBC_Sched_op ((char *) redbuf - (intptr_t) handle->tmpbuf, true, sendbuf, false, 0, true, count,
-                                datatype, op, schedule, true);
-          }
+        if (firstred && MPI_IN_PLACE != sendbuf) {
```

Comment:
i think, ( MPI_IN_PLACE != sendbuf ) is always true.
Because sendbuf changes recvbuf (sendbuf = recvbuf in nbc_internal.h:602)
when NBC_IN_PLACE() is called and if (sendbuf == MPI_IN_PLACE) is true.
So, (firstred && !inplace) is correct, isn't it?